aptos-move/aptos-aggregator/src/delayed_field_extension.rs

```diff
@@ -2,347 +2,120 @@
-        DelayedFieldsSpeculativeError, PanicError, PanicOr, ReadPosition, SnapshotToStringFormula,
+        code_invariant_error, expect_ok, DelayedFieldID, DelayedFieldValue,
+        DelayedFieldsSpeculativeError, PanicOr, ReadPosition, SnapshotToStringFormula,
         SnapshotValue,
     },
 };
-use aptos_types::{state_store::state_key::StateKey, vm_status::StatusCode};
-use claims::assert_matches;
-use move_binary_format::errors::{PartialVMError, PartialVMResult};
-use std::collections::{btree_map::Entry, BTreeMap, BTreeSet};
-
-/// Describes how the `speculative_start_value` in `AggregatorState` was obtained.
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum SpeculativeStartValue {
-    // The speculative_start_value is not yet initialized
-    Unset,
-    // The speculative_start_value was obtained by reading
-    // the last committed value of the aggregator from MVHashmap.
-    // WARNING: any use of this value should be captured as a restriction
-    // in the change set, as value received here is not track as part of the
-    // read conflict!!
-    // Only current restriction is DeltaHistory, and only correct usage is
-    // that can be returned to the caller is via try_add/try_sub methods!
-    LastCommittedValue(u128),
-    // The speculative_start_value was obtained by performing a read
-    // procedure on the aggregator, which involves aggregating deltas
-    // present at the read time
-    AggregatedValue(u128),
-}
-
-impl SpeculativeStartValue {
-    // WARNING: any use of this value should be captured as a restriction
-    // in the change set, as value received here is not track as part of the
-    // read conflict!!
-    // Only current restriction is DeltaHistory, and only correct usage is
-    // that can be returned to the caller is via try_add/try_sub methods!
-    pub fn get_any_value(&self) -> Result<u128, PanicError> {
-        match self {
-            SpeculativeStartValue::Unset => Err(code_invariant_error(
-                "Tried calling get_any_value on Unset speculative value",
-            )),
-            SpeculativeStartValue::LastCommittedValue(value) => Ok(*value),
-            SpeculativeStartValue::AggregatedValue(value) => Ok(*value),
-        }
-    }
-
-    pub fn get_value_for_read(&self) -> Result<u128, PanicError> {
-        match self {
-            SpeculativeStartValue::Unset => Err(code_invariant_error(
-                "Tried calling get_value_for_read on Unset speculative value",
-            )),
-            SpeculativeStartValue::LastCommittedValue(_) => Err(code_invariant_error(
-                "Tried calling get_value_for_read on LastCommittedValue speculative value",
-            )),
-            SpeculativeStartValue::AggregatedValue(value) => Ok(*value),
-        }
-    }
-}
-
-/// Describes the state of each aggregator instance.
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum AggregatorState {
```

Comment:
I'm a little bit confused here. Why are SpeculativeStartValue, AggregatorState, AggregatorSnapshot, AggregatorSnapshotState removed? I don't seem in the imports as well?

Reply:
yes your eyes are not tricking you - they are removed :) 

SpeculativeStartValue is not stored/kept in the extension any more. CapturedReads caches it instead - to make sure we don't go twice to disk. This is same place where caching happens for resources - so is well aligned with the rest of the code.

after removing SpeculativeStartValue, what is left is identical to what goes into the change set (if you remember you tried reusing the same class, but we ended up having two).

So - because it goes into the VMChangeSet, I call it DelayedChange (and not state)